File: projects/hyperion/src/core/math/plane.cpp
#include "hyppch.hpp"

#include "hyperion/core/math/plane.hpp"

namespace Hyperion {

    Plane::Plane()
        : normal(Vec3::Up()), distance(0) { }

    Plane::Plane(Vec3 normal, float32 distance)
        : normal(normal.Normalized()), distance(distance) { }

    Plane::Plane(Vec3 normal, Vec3 point) {
        this->normal = normal.Normalized();
        distance = -this->normal.Dot(point);
    }

    Plane::Plane(Vec3 a, Vec3 b, Vec3 c) {
        normal = Vec3::Cross(b - a, c - a).Normalized();
        distance = -this->normal.Dot(a);
    }

    void Plane::Flip() {
        normal = -normal;
        distance = -distance;
    }

    Vec3 Plane::GetClosestPoint(Vec3 point) const {
        float32 d = normal.Dot(point) + distance;
        return point - normal * d;
    }

    float32 Plane::GetDistanceToPoint(Vec3 point) const {
        return normal.Dot(point) + distance;
    }

    bool Plane::Intersects(Ray ray, float32 &hit_distance) const {
        float32 angle = ray.direction.Dot(normal);
        float32 origin_to_plane = -ray.origin.Dot(normal) - this->distance;
        bool hit;
        if (angle == 0.0f) {
            hit_distance = 0.0f;
            hit = false;
        } else {
            float32 dis = origin_to_plane / angle;
            hit_distance = dis;
            hit = dis > 0.0f;
        }
        return hit;
    }

    String Plane::ToString() const {
        return StringUtils::Format("(Normal: {}, Distance: {})", normal.ToString(), distance);
    }

    bool Plane::operator==(const Plane &other) const {
        return normal == other.normal && distance == other.distance;
    }

    bool Plane::operator!=(const Plane &other) const {
        return !(*this == other);
    }

}

File: projects/hyperion-managed/Hyperion.Core/Math/Plane.cs
using System;
using System.Runtime.CompilerServices;

namespace Hyperion {
    public struct Plane : IEquatable<Plane> {
        public Vector3 Normal;
        public float Distance;

        public Plane(Vector3 normal, float distance) {
            Normal = normal;
            Distance = distance;
        }

        public Plane(Vector3 normal, Vector3 point) {
            Normal = normal.Normalized;
            Distance = -Vector3.Dot(Normal, point);
        }

        public Plane(Vector3 a, Vector3 b, Vector3 c) {
            Normal = Vector3.Cross(b - a, c - a).Normalized;
            Distance = -Vector3.Dot(Normal, a);
        }

        public void Flip() {
            Normal = -Normal;
            Distance = -Distance;
        }

        public Vector3 GetClosestPoint(Vector3 point) {
            float d = Vector3.Dot(Normal, point) + Distance;
            return point - Normal * d;
        }

        public float GetDistanceToPoint(Vector3 point) {
            return Vector3.Dot(Normal, point) + Distance;
        }

        public bool Intersects(Ray ray, out float distance) {
            float angle = Vector3.Dot(ray.Direction, Normal);
            float originToPlane = -Vector3.Dot(ray.Origin, Normal) - Distance;
            bool hit;
            if (angle == 0.0f) {
                distance = 0.0f;
                hit = false;
            } else {
                float dis = originToPlane / angle;
                distance = dis;
                hit = dis > 0.0f;
            }
            return hit;
        }

        public override string ToString() {
            return $"(Normal: {Normal}, Distance: {Distance})";
        }

        public override bool Equals(object obj) {
            if (obj is Plane other) {
                return this == other;
            } else {
                return false;
            }
        }

        public override int GetHashCode() {
            return Normal.GetHashCode() ^ Distance.GetHashCode() << 2;
        }

        public bool Equals(Plane other) {
            return this == other;
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static bool operator ==(Plane lhs, Plane rhs) {
            return lhs.Normal == rhs.Normal && lhs.Distance == rhs.Distance;
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static bool operator !=(Plane lhs, Plane rhs) {
            return !(lhs == rhs);
        }
    }
}
